File: langchain_agent/mcp_agent.py
#!/usr/bin/env python3
"""
MCP Agent CLI - Interactive terminal agent connected to MCP servers.

Usage:
    python mcp_agent.py --mcp-url http://localhost:8000/mcp
    python mcp_agent.py --mcp-url http://localhost:8000/mcp --auth-header "Bearer token"
"""

import argparse
import asyncio
import os
import sys

from langchain_core.messages import AIMessage, HumanMessage, ToolMessage
from rich.console import Console
from rich.live import Live
from rich.markdown import Markdown
from rich.panel import Panel
from rich.spinner import Spinner

from .agent.graph import AgentConfig, run_agent_session
from .auth.oauth import CallbackServer, create_oauth_provider


console = Console()


def parse_args() -> argparse.Namespace:
    """Parse command line arguments."""
    parser = argparse.ArgumentParser(
        description="Interactive MCP Agent - Chat with AI using MCP server tools",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  # Basic usage (will auto-handle OAuth if server requires it)
  python mcp_agent.py --mcp-url http://localhost:8000/mcp

  # With static auth header
  python mcp_agent.py --mcp-url http://localhost:8000/mcp --auth-header "Bearer my-token"

Environment Variables:
  ANTHROPIC_API_KEY    API key for Anthropic (required if using Anthropic provider)
  OPENAI_API_KEY       API key for OpenAI (required if using OpenAI provider)
  MCP_AGENT_PROVIDER   Default LLM provider (anthropic or openai)
  MCP_AGENT_MODEL      Default LLM model
        """,
    )

    # Required arguments
    parser.add_argument(
        "--mcp-url",
        required=True,
        help="MCP server URL (e.g., http://localhost:8000/mcp)",
    )

    # Authentication options
    auth_group = parser.add_argument_group("Authentication")
    auth_group.add_argument(
        "--auth-header",
        help="Authorization header value (e.g., 'Bearer token123')",
    )
    auth_group.add_argument(
        "--no-oauth",
        action="store_true",
        help="Disable automatic OAuth flow",
    )

    # LLM options
    llm_group = parser.add_argument_group("LLM Configuration")
    llm_group.add_argument(
        "--provider",
        choices=["anthropic", "openai"],
        default=os.environ.get("MCP_AGENT_PROVIDER", "anthropic"),
        help="LLM provider (default: anthropic or MCP_AGENT_PROVIDER env var)",
    )
    llm_group.add_argument(
        "--model",
        default=os.environ.get("MCP_AGENT_MODEL"),
        help="LLM model name (default: provider-specific default or MCP_AGENT_MODEL env var)",
    )

    return parser.parse_args()


def print_banner() -> None:
    """Print the application banner."""
    console.print()
    console.print("╔══════════════════════════════════════════════════╗", style="bold blue")
    console.print("║  🤖 MCP Agent - Interactive Terminal Chat       ║", style="bold blue")
    console.print("╚══════════════════════════════════════════════════╝", style="bold blue")
    console.print()


def format_tool_call(tool_name: str, tool_args: dict) -> Panel:
    """Format a tool call for display."""
    args_str = "\n".join(f"  {k}: {v}" for k, v in tool_args.items())
    content = f"[bold cyan]{tool_name}[/bold cyan]\n{args_str}" if args_str else f"[bold cyan]{tool_name}[/bold cyan]"
    return Panel(content, title="🔧 Tool Call", border_style="cyan", expand=False)


def format_tool_result(result: str) -> Panel:
    """Format a tool result for display."""
    # Truncate very long results
    if len(result) > 500:
        result = result[:500] + "...[truncated]"
    return Panel(result, title="📋 Result", border_style="green", expand=False)


def display_agent_response(result: dict) -> str:
    """
    Display the agent response with tool calls and final answer.
    Returns the final text response.
    """
    messages = result.get("messages", [])
    final_response = ""
    
    for message in messages:
        if isinstance(message, AIMessage):
            # Check for tool calls
            if hasattr(message, "tool_calls") and message.tool_calls:
                for tool_call in message.tool_calls:
                    console.print(format_tool_call(
                        tool_call.get("name", "unknown"),
                        tool_call.get("args", {})
                    ))
            
            # Display text content if present
            if message.content:
                if isinstance(message.content, str):
                    final_response = message.content
                elif isinstance(message.content, list):
                    # Handle list of content blocks
                    text_parts = []
                    for block in message.content:
                        if isinstance(block, dict) and block.get("type") == "text":
                            text_parts.append(block.get("text", ""))
                        elif isinstance(block, str):
                            text_parts.append(block)
                    final_response = "".join(text_parts)
        
        elif isinstance(message, ToolMessage):
            # Display tool results
            content = message.content
            if isinstance(content, str):
                console.print(format_tool_result(content))
            else:
                console.print(format_tool_result(str(content)))
    
    # Display the final response as markdown
    if final_response:
        console.print(Markdown(final_response))
    
    return final_response


async def run_agent_query(agent, messages: list[dict]) -> dict:
    """Run the agent with the given messages and return the result."""
    with Live(Spinner("dots", text="Thinking..."), console=console, refresh_per_second=10):
        result = await agent.ainvoke({"messages": messages})
    return result


async def chat_loop(agent) -> None:
    """Run the interactive chat loop."""
    messages = []
    
    console.print(
        "[dim]Type your message and press Enter. Use Ctrl+C or /quit to exit.[/dim]\n"
    )

    while True:
        try:
            # Get user input
            user_input = console.input("[bold green]You:[/bold green] ")
            
            if not user_input.strip():
                continue
            
            # Handle special commands
            if user_input.lower() in ["/quit", "/exit", "/q"]:
                console.print("\n[dim]Goodbye! 👋[/dim]")
                break
            
            if user_input.lower() == "/clear":
                messages = []
                console.clear()
                print_banner()
                console.print("[dim]Conversation cleared.[/dim]\n")
                continue
            
            if user_input.lower() == "/help":
                console.print(Panel(
                    "[bold]/quit, /exit, /q[/bold] - Exit the chat\n"
                    "[bold]/clear[/bold] - Clear conversation history\n"
                    "[bold]/help[/bold] - Show this help message",
                    title="Commands",
                    border_style="blue",
                ))
                continue

            # Add user message to history
            messages.append({"role": "user", "content": user_input})
            
            console.print()
            console.print("[bold blue]Assistant:[/bold blue]")
            
            # Run agent and get response
            result = await run_agent_query(agent, messages)
            
            # Display the response (including tool calls)
            response = display_agent_response(result)
            
            # Add assistant response to history for context
            if response:
                messages.append({"role": "assistant", "content": response})
            
            console.print()

        except KeyboardInterrupt:
            console.print("\n\n[dim]Goodbye! 👋[/dim]")
            break
        except Exception as e:
            console.print(f"\n[red]Error: {e}[/red]\n")
            import traceback
            traceback.print_exc()


async def main() -> None:
    """Main entry point."""
    args = parse_args()
    
    print_banner()
    
    console.print(f"[dim]Connecting to MCP server: {args.mcp_url}[/dim]")
    
    # Determine base URL (remove /mcp suffix for OAuth)
    base_url = args.mcp_url
    if base_url.endswith("/mcp"):
        base_url = base_url[:-4]
    
    # Set up authentication
    oauth_provider = None
    callback_server = None
    
    if not args.auth_header and not args.no_oauth:
        # Set up OAuth flow
        callback_server = CallbackServer(port=3030)
        callback_server.start()
        oauth_provider = create_oauth_provider(base_url, callback_server)

    # Create agent configuration
    config = AgentConfig(
        mcp_url=args.mcp_url,
        provider=args.provider,
        model=args.model,
        auth_header=args.auth_header,
        oauth_provider=oauth_provider,
    )

    console.print(f"[dim]Using LLM: {args.provider}/{config.model}[/dim]")
    console.print()

    try:
        async def on_agent_ready(agent, client, tools):
            """Called when agent is ready."""
            console.print()
            await chat_loop(agent)

        # Run the agent session
        await run_agent_session(config, on_agent_ready)

    except ValueError as e:
        console.print(f"[red]Configuration error: {e}[/red]")
        sys.exit(1)
    except Exception as e:
        console.print(f"[red]Error: {e}[/red]")
        import traceback
        traceback.print_exc()
        sys.exit(1)
    finally:
        if callback_server:
            callback_server.stop()


if __name__ == "__main__":
    asyncio.run(main())

File: claude-agent-sdk-typescript/src/agent/mcp-agent.ts
/**
 * MCP Agent configuration and session management using Claude Agent SDK.
 */

import { query, type Options, type McpHttpServerConfig, type SDKMessage } from "@anthropic-ai/claude-agent-sdk";

export interface AgentConfig {
  mcpUrl: string;
  authToken?: string;
  model?: string;
}

/**
 * Create MCP server configuration for the Claude Agent SDK
 */
export function createMcpServerConfig(config: AgentConfig): Options["mcpServers"] {
  const serverConfig: McpHttpServerConfig = {
    type: "http",
    url: config.mcpUrl,
  };

  // Add authorization header if token provided
  if (config.authToken) {
    serverConfig.headers = {
      Authorization: config.authToken.startsWith("Bearer ")
        ? config.authToken
        : `Bearer ${config.authToken}`,
    };
  }

  return {
    "mcp-server": serverConfig,
  };
}

/**
 * Extract text content from an SDK message
 */
function extractMessageContent(message: SDKMessage): string {
  if (message.type === "assistant") {
    // SDKAssistantMessage has a `message` property containing the API response
    const apiMessage = message.message;
    if (apiMessage && apiMessage.content) {
      if (typeof apiMessage.content === "string") {
        return apiMessage.content;
      }
      if (Array.isArray(apiMessage.content)) {
        const textBlocks = apiMessage.content.filter(
          (block: { type: string }): block is { type: "text"; text: string } =>
            block.type === "text"
        );
        return textBlocks.map((block: { type: "text"; text: string }) => block.text).join("");
      }
    }
  } else if (message.type === "result") {
    // SDKResultMessage has a `result` property only on success subtype
    if (message.subtype === "success") {
      return message.result || "";
    }
  }
  return "";
}

/**
 * Run a single query against the MCP server
 */
export async function runQuery(
  prompt: string,
  config: AgentConfig,
  conversationHistory: Array<{ role: string; content: string }> = []
): Promise<string> {
  const mcpServers = createMcpServerConfig(config);

  // Build the full prompt with conversation history
  let fullPrompt = prompt;
  if (conversationHistory.length > 0) {
    const historyText = conversationHistory
      .map((msg) => `${msg.role}: ${msg.content}`)
      .join("\n");
    fullPrompt = `Previous conversation:\n${historyText}\n\nUser: ${prompt}`;
  }

  const options: Options = {
    mcpServers,
    model: config.model || "claude-sonnet-4-20250514",
    permissionMode: "acceptEdits",
    // Allow all tools from the MCP server (mcp__<server-name>__<tool-name>)
    allowedTools: ["mcp__mcp-server__*"],
  };

  let response = "";

  try {
    const messages = query({ prompt: fullPrompt, options });

    for await (const message of messages) {
      // Log system init message to show MCP server status
      if (message.type === "system" && "subtype" in message && message.subtype === "init") {
        const initMessage = message as { mcp_servers?: Array<{ name: string; status: string }>; tools?: string[] };
        if (initMessage.mcp_servers) {
          console.log("\n[MCP Servers]");
          for (const server of initMessage.mcp_servers) {
            const statusIcon = server.status === "connected" ? "✓" : "✗";
            console.log(`  ${statusIcon} ${server.name}: ${server.status}`);
          }
        }
        if (initMessage.tools) {
          const mcpTools = initMessage.tools.filter((t: string) => t.startsWith("mcp__"));
          if (mcpTools.length > 0) {
            console.log("\n[MCP Tools Available]");
            for (const tool of mcpTools) {
              console.log(`  • ${tool}`);
            }
          }
        }
        console.log();
      }

      if (message.type === "assistant") {
        const content = extractMessageContent(message);
        if (content) {
          response = content;
        }

        // Log tool usage if present
        const apiMessage = message.message;
        if (apiMessage && apiMessage.content && Array.isArray(apiMessage.content)) {
          for (const block of apiMessage.content) {
            if (block.type === "tool_use") {
              console.log(`\n[Tool Call] ${block.name}`);
              console.log(`  Args: ${JSON.stringify(block.input, null, 2)}`);
            }
          }
        }
      } else if (message.type === "result") {
        // Final result message
        const resultContent = extractMessageContent(message);
        if (resultContent) {
          response = resultContent;
        }
      }
    }
  } catch (error) {
    if (error instanceof Error) {
      throw new Error(`Agent query failed: ${error.message}`);
    }
    throw error;
  }

  return response;
}

/**
 * Stream a query and yield responses as they come
 */
export async function* streamQuery(
  prompt: string,
  config: AgentConfig
): AsyncGenerator<{ type: string; content: string }> {
  const mcpServers = createMcpServerConfig(config);

  const options: Options = {
    mcpServers,
    model: config.model || "claude-sonnet-4-20250514",
    permissionMode: "acceptEdits",
    // Allow all tools from the MCP server (mcp__<server-name>__<tool-name>)
    allowedTools: ["mcp__mcp-server__*"],
  };

  const messages = query({ prompt, options });

  for await (const message of messages) {
    if (message.type === "assistant") {
      const content = extractMessageContent(message);
      if (content) {
        yield { type: "assistant", content };
      }

      // Yield tool calls
      const apiMessage = message.message;
      if (apiMessage && apiMessage.content && Array.isArray(apiMessage.content)) {
        for (const block of apiMessage.content) {
          if (block.type === "tool_use") {
            yield {
              type: "tool_call",
              content: `${block.name}: ${JSON.stringify(block.input)}`,
            };
          }
        }
      }
    } else if (message.type === "result") {
      yield {
        type: "result",
        content: extractMessageContent(message),
      };
    }
  }
}
